_posts/2019/2019-06-20-managing-lets-encrypt-certificates-in-vault.md

```diff
@@ -0,0 +1,221 @@
+tightly controls access to tokens, passwords, certificates, API keys, and other
+secrets in modern computing.[^vault]
+
+To avoid depending on a single server, we containerize the complete process.
+Since the state is stored in Vault, we are able to run the certificate creation
+and renewal process anywhere as long as there is Vault access.
+
+A testing setup can be created by using the files provided in [this
+Gist](https://gist.github.com/ikoutras-epages/b4780862a93315aac1a971754ca50eab){:target="_blank"}.
+In the following sections we describe how they are supposed to work and how one
+issues and renews certificates.
+
+## Get Vault ready
+
+Before running anything we need a Vault instance and a proper token to access
+it. For the sake of this post, let's run Vault locally, inside a Docker
+container:
+
+```bash
+$ docker network create certbot-vault-net
+$ docker run --cap-add=IPC_LOCK -d --name=dev-vault \
+  --network certbot-vault-net \
+  vault
+```
+
+The first command creates a Docker network, so that the Certbot container can
+access the Vault. The second creates a Vault container based on the [official
+Vault image](https://hub.docker.com/_/vault/){:target="_blank"} (version 1.1.3 was the latest
+version we tested). Be careful, this Vault instance is running on "dev mode",
+which means that every data will be lost on container stop.
+
+Let's grab the root token to use it for Certbot:
+
+```bash
+export VAULT_TOKEN=$(docker logs dev-vault 2>&1 | grep 'Root Token' | cut -d ' ' -f3)
+```
+
+*Note: On a production environment it is advised to follow the principle of
+least privilege, i.e., use multiple tokens with different policies, depending
+on the process. When initializing Certbot data and/or creating new
+certificates, we need write permissions on everything, while when performing
+certificate renewal, we need read access on most Vault paths and only update
+permission on the certificates path.*
+
+## Create a Docker image for Certbot and Vault
+
+Let's embed the Vault client on an official Certbot Docker image. As
+aforementioned, we are going to use DNS-01 challenges to avoid exposing any
+network ports and re-directing HTTP(S) traffic to this container. More
+specifically, we are going to use the DNSimple plug-in as seen in the
+[Dockerfile](https://gist.github.com/ikoutras-epages/b4780862a93315aac1a971754ca50eab#file-dockerfile){:target="_blank"}.
+
+The RUN argument is similar to the [official's Vault Docker
+image](https://github.com/hashicorp/docker-vault/blob/2b7561b55940c35e412c914083e7dd40d21e9193/0.X/Dockerfile#L12-L46){:target="_blank"}
+with the removal of the `ca-certificates` package which is available from the
+base image and the addition of the `curl` and `jq` packages that we are going
+to need later on.
+
+Let's build this and tag it as `certbot-vault` by running the following on the
+same directory of our `Dockerfile`:
+
+```bash
+docker build -t certbot-vault .
```

Comment:
Is the dot correct here after certbot-vault?

Reply:
Yes, it is the directory into which Docker should look in order to find the Dockerfile.